### ISSUE TITLE
Add support for `--terragrunt-source-map` as a more consistent, predictable way to override source URLs

### ISSUE THREAD

**Issue by yorinasub17** [gruntwork-io/terragrunt#1138]
Terragrunt's `--terragrunt-source` parameter, which is designed to make it easy to override `source` URLs in `terragrunt.hcl`, has a couple issues:

1. If you run `terragrunt apply --terragrunt-source /foo/bar` in folder `yyy`, as of now, the *only* thing it does is override the `source` URL in `yyy/terragrunt.hcl` to the exact value `/foo/bar`. However, if you run `terragrunt apply-all --terragrunt-source /foo/bar` in folder `yyy`, it overrides the `source` value of each child module within `yyy/sub/path` to something like `/foo/bar/sub/path`. This inconsistency can be confusing and only gets worse when working with `read_terragrunt_config`, where we have to parse another `terragrunt.hcl` and decide whether to override its `source` URL or not.
1. `--terragrunt-source` assumes all your modules are coming from the *same* root repo. You won't be able to do an `apply-all` if you're pulling from multiple repos (e.g., your own infra modules, Terraform Registry, Gruntwork repos, etc). 

Perhaps a better solution to this is to support some sort of mapping:

```
terragrunt apply --terragrunt-source-map source:dest
```

The `--terragrunt-source-map` param would replace any `source` URL that starts with `source` to instead start with `dest`. 

For example:

```
terragrunt apply --terragrunt-source-map github.com/org/modules.git:/local/path/to/modules
```

The above would replace `source = "github.com/org/modules.git//xxx"` with `/local/path/to/modules//xxx` regardless of whether you were running `apply`, or `apply-all`, or using a `dependency`. This would give us a consistent behavior across all commands _and_ support mapping multiple repos by specifying the param multiple times:

```
terragrunt apply \
  --terragrunt-source-map github.com/org/modules.git:/local/path/to/modules \
  --terragrunt-source-map github.com/org/another-repo.git:/local/path/to/another-repo
```

We could also consider allowing users to specify this mapping in a file:

```
terragrunt apply --terragrunt-source-map mapping.hcl
```

Where `mapping.hcl` contains:

```hcl
mappings = {
  "github.com/org/modules.git"      = "/local/path/to/modules"
  "github.com/org/another-repo.git" = "/local/path/to/another-repo"  
}
```

See https://github.com/gruntwork-io/terragrunt/pull/1134#discussion_r407167726 for context for this issue.

**Comment by yorinasub17** [comment]
Core feature is done, but still missing the following:

- [ ] Environment variable configuration
- [ ] file based configuration

**Comment by yorinasub17** [comment]
Env var config PR: https://github.com/gruntwork-io/terragrunt/pull/1676

**Comment by queglay** [comment]
I am wondering, can this handle the immutable ref suffix if used?
https://github.com/gruntwork-io/terragrunt/issues/1627#issuecomment-847082445

in the example linked:
```
source = "github.com/firehawkvfx/firehawk-main.git//modules/vault?ref=v0.0.20"
source = "github.com/firehawkvfx/firehawk-main.git//modules/vault-kv-init?ref=v0.0.20"
source = "github.com/firehawkvfx/firehawk-main.git//modules/vault-policies?ref=v0.0.20"
```
We would want to override all these to point at another path to test before the next release.

**Comment by yorinasub17** [comment]
Yes this should work with `--terragrunt-source-map github.com/firehawkvfx/firehawk-main.git=/path/to/local/modules`. The source map feature ignores and drops query parameters when doing the mapping.

**Comment by queglay** [comment]
Thank you for clarifying.

**Comment by ppozeti** [comment]
First of all this is a great feature! I've got a lot of nested modules and it is a pain to test them.

Sadly I got no luck configuring that in Windows. I tried to run like `--terragrunt-source-map git::https://azuredevopsurl/_git/=C:\IaC` and it fails with "Did not find any Terraform files (*.tf)".

I compared debug logs with/without the --terragrunt-source-map argument and it seems that the "WorkingDir" variable gets a non hashed value when using --terragrunt-source-map with C:\ in the path.

Have anyone being able to use this with Windows?

**Comment by davidwneary** [comment]
> The source map feature ignores and drops query parameters when doing the mapping.

Are there any plans to allow this behaviour to be configurable? e.g. `--preserve-query-parameters` or equivalent?

My use case is that I define the paths to my modules using something like 
```
git::ssh://git@gitlab.com/company/project.git?ref=1.0
```
This means a developer can check the project out and run a plan without any faffing with credentials or personal access tokens.

However, my CI runner doesn't have an SSH key, so would ideally use a path like

```
git::https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com/company/project.git?ref=1.0
```
I tried using TERRAFORM_SOURCE_MAP for this but unfortunately, because the query parameters are ignored, it will always point to the master branch rather than the ref specified.

Cheers!

**Comment by angeloskaltsikis** [comment]
> > The source map feature ignores and drops query parameters when doing the mapping.
> 
> Are there any plans to allow this behaviour to be configurable? e.g. `--preserve-query-parameters` or equivalent?
> 
> My use case is that I define the paths to my modules using something like
> 
> ```
> git::ssh://git@gitlab.com/company/project.git?ref=1.0
> ```
> 
> This means a developer can check the project out and run a plan without any faffing with credentials or personal access tokens.
> 
> However, my CI runner doesn't have an SSH key, so would ideally use a path like
> 
> ```
> git::https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com/company/project.git?ref=1.0
> ```
> 
> I tried using TERRAFORM_SOURCE_MAP for this but unfortunately, because the query parameters are ignored, it will always point to the master branch rather than the ref specified.
> 
> Cheers!

Can you try to add to your CI `.gitconfig` the following contents like Atlantis does 
https://github.com/runatlantis/helm-charts/blob/main/charts/atlantis/values.yaml#L53 ?
I think this will solve your issue.

**Comment by davidwneary** [comment]
@angeloskaltsikis amazing! Thanks so much for the quick response, this did just the trick. I added the following as a pre-script:

```
git config --global url."https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com".insteadOf ssh://git@gitlab.com
```
And now no need to use TERRAFORM_SOURCE_MAP at all in the CI. Thanks a lot!

**Comment by angeloskaltsikis** [comment]
> @angeloskaltsikis amazing! Thanks so much for the quick response, this did just the trick. I added the following as a pre-script:
> 
> ```
> git config --global url."https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com".insteadOf ssh://git@gitlab.com
> ```
> 
> And now no need to use TERRAFORM_SOURCE_MAP at all in the CI. Thanks a lot!

That's great to hear. I would be glad if you can add me in LinkedIn to discuss further about your CI and how do you use it :-)

**Comment by vyom-soft** [comment]
> > @angeloskaltsikis amazing! Thanks so much for the quick response, this did just the trick. I added the following as a pre-script:
> > ```
> > git config --global url."https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com".insteadOf ssh://git@gitlab.com
> > ```
   
> > And now no need to use TERRAFORM_SOURCE_MAP at all in the CI. Thanks a lot!
> 
> That's great to hear. I would be glad if you can add me in LinkedIn to discuss further about your CI and how do you use it :-)

Thanks, It is works for the Main branch. But when I create another branch and run the pipeline I see the following err.
Can you pls. suggest something.

```
$ terragrunt init -upgrade
time=2024-02-06T10:38:06Z level=error msg=1 error occurred:
	* error downloading 'ssh://git@gitlab.bare.mycompany.org/edp/service/mynet/iam1/terraform-modules.git?ref=HEAD': /usr/bin/git exited with 128: Cloning into '/builds/edp/service/mynet/iam1/keycloak-terragrunt/environments/test_env/.terragrunt-cache/ZZZZ-u3X9KRbErgeRuqI-FaZT0/q-vjwbnLERG79bd2l12h4OVvHZA'...
remote: HTTP Basic: Access denied. The provided password or token is incorrect or your account has 2FA enabled and you must use a personal access token instead of a password. See https://gitlab.bare.pandrosion.org/help/topics/git/troubleshooting_git#error-on-git-fetch-http-basic-access-denied
fatal: Authentication failed for 'https://gitlab.bare.mycompany.org/edp/service/mynet/iam1/terraform-modules.git/'
time=20[24](https://gitlab.bare.mycompany.org/edp/service/mynet/iam1/keycloak-terragrunt/-/jobs/673296#L24)-02-06T10:38:06Z level=error msg=Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

**Comment by yhakbar** [comment]
This is available in the CLI:
https://terragrunt.gruntwork.io/docs/reference/cli-options/#source-map
